### DISCUSSION TITLE
Connecting to the inverter -wiki: Sample case Using a dedicated WLAN Bridge: TP-link TL-WR802N

### DISCUSSION THREAD

**Discussion by MEKadan** [wlcrs/huawei_solar#83] [Setting up the integration] [answered]
From wiki mentioned above:

How exactly you configure the router?

-  Setting the client mode is OK
- I get an IP for TP-Link in my 192.168.1.x network after setting it into client mode.
- In TP-Link I can feed in the parameters for the inverter ap ( host to be connected in client mode, that is SSID & PW)

Now, I have not set the DHCP in TP-Link, i'm in undertstanding it is not used in client mode.
I have installed the FusionSolar app, as long as it is needed, to get the initial connection and I can get to commissioning etc. However, I have not set the router connections, because the whole point of using this dedicated wlan bridge is to isolate my home network from these chinese systems.

Also the static route has been set in mys asus router so tha network/host is 192.168.200.0, netmask 255.255.255.0 and gateway is the address of TP-Link

I can access TP-Link OK, but cannot see anything behind it (well, there should be only inverter at 192.168.200.2), I cannot ping it or traceroute it.

Should it be answering at this stage or are there still things to be done?

inverter model is SUN2000MA-4KTL-M1 firmware is V100R001C00SPC141
Edit: About dongle (Sdongle-05-FE,  "WLAN-FE Smart dongle") sofware version I have no idea. It just is not anywhere to be found/ updated. In device upgrades I have only inverter and MBUS, not even a half word of dongle

**Comment by Karakth** [comment]
Hey,

I haven't managed to get it to work either but some pointers:
1) The network should be 192.168.200.0/24
2) Inverter should be at 192.168.200.1

Hope this helps. I haven't managed to get it to work via unifi. Good luck!

**Comment by wlcrs** [comment, marked as answer]
Can you draw a diagram of how everything is connected with each other? How is the TP-Link router connected to your home network? What devices have which IP(s)? Where did you set the routes to 192.168.200.0/24 network?

**Reply by MEKadan** [reply]
Hi @wlcrs , if your question was targeted to me, I managed to get it going as described in this thread https://community.home-assistant.io/t/integration-solar-inverter-huawei-2000l/132350/1864 starting somewhere near post 1850 or so. Long discussion. 

Key thing was to replace TL-WR802N's own software with OpenWRT. One must also notice that theere was a bug in the recent versions of it. which killed the firewall (and thus all the rules needed to let packets through)  Latest versions should be OK, I have 22.03.0-rc4 if i remeber correctly. This, and probably newer ones work. However, there are also plenty of guides around the net usin IPTABLES to configure firewall rules, but IPTABLES is now replaced with nftables, which causes some extra headache. This is also handled in the thread i linked above. The thread should have all the bits and pieces to get it to work with TL-WR802N with OpenWRT.

I'm a bit reserved  at this time of the year, and most of the time out of office, or should I says out of home... so I don't have possibility to make more detailed story in a near future, chairman of  the board (meaning my wife) is keeping me occuppied in holiday season :) :)

**Reply by wlcrs** [reply]
I rewrote the [OpenWRT instructions on the Wiki](https://github.com/wlcrs/huawei_solar/wiki/Connecting-to-the-inverter#using-openwrt) a bit to use a DNAT rule instead of setting up proper routing to the 192.168.200.0 subnet, as this is easier for people to execute that have limited computer networking knowledge:

**Reply by henfri** [reply]
Hello,

the instructions in the Wiki did not work for me.
I changed three things (unfortunately at the same time) before it started working:
```
config forwarding
        option dest 'sun'
        option src 'lan'

config forwarding
        option dest 'lan'
        option src 'sun'
```

And this:
https://user-images.githubusercontent.com/83491212/183281785-2c9ad9c1-2939-4cbf-8267-cdfb93f1d296.png

And this:
https://user-images.githubusercontent.com/83491212/183281779-a755a952-005f-413c-a6e9-a989f6b00f30.png

**Comment by yousaf465** [comment]
> Hi @wlcrs , if your question was targeted to me, I managed to get it going as described in this thread https://community.home-assistant.io/t/integration-solar-inverter-huawei-2000l/132350/1864 starting somewhere near post 1850 or so. Long discussion.
> 
> Key thing was to replace TL-WR802N's own software with OpenWRT. One must also notice that theere was a bug in the recent versions of it. which killed the firewall (and thus all the rules needed to let packets through) Latest versions should be OK, I have 22.03.0-rc4 if i remeber correctly. This, and probably newer ones work. However, there are also plenty of guides around the net usin IPTABLES to configure firewall rules, but IPTABLES is now replaced with nftables, which causes some extra headache. This is also handled in the thread i linked above. The thread should have all the bits and pieces to get it to work with TL-WR802N with OpenWRT.
> 
> I'm a bit reserved at this time of the year, and most of the time out of office, or should I says out of home... so I don't have possibility to make more detailed story in a near future, chairman of the board (meaning my wife) is keeping me occuppied in holiday season :) :)

https://forum.openwrt.org/t/connect-to-huwaei-sun2000-inverter-using-fiber-modem/104635/23?u=koa 

I summarized the details in one post mentioned above. 

<details><summary>CLICK ME</summary>
So to summarize, I did this. Note Huawei inverter is on another subnet that is 192.168.200.0.  
![image](https://user-images.githubusercontent.com/83491212/183281643-390822a9-834b-4fd3-ace4-ba65e7dfd930.png)
1. Flash Openwrt on the router
2. Setup your router as a dumb AP. https://openwrt.org/docs/guide-user/network/wifi/dumbap
3. Connect it to your ISP router via ethernet. This will change the IP to something in the router’s subnet range. e.g if your ISP’s router Ip is 192.168.1.1 then your OpenWrt's router IP will be 192.168.1.24 (anything in between 192.168.1.2 to 192.168.1.100)
4. Connect to your Huwaei Sun2000 inverter’s wifi as a client via 2.4Ghz radio.
  * Network – Wireless
  * click scan
![image](https://user-images.githubusercontent.com/83491212/183281654-eb4aa929-b5a3-47c7-b339-8ebce54ac582.png)

* connect to SUN2000-HVxxxxxxxx
* Use password “Changeme”
* Wireless security WPA2-PSK
![image](https://user-images.githubusercontent.com/83491212/183281667-a3c0e20f-6d56-4008-ad05-0c1249fe0732.png)
Check the connection between OpenWrt and router by ping 192.168.200.1 (in most cases but in some cases it will be 192.168.199.1)

* Network – Diagnostics
![image](https://user-images.githubusercontent.com/83491212/183281683-6f49c00f-dc00-4ddb-8448-32a650e9204b.png)

------------------------------------------------------------------------------------------------
**Step 2** 
This step is unique for each user, depending on your ISP modem/ router. 
Set up a static route in ISP router/modem/ 
e.g 
**On F@st 5366**

1. Open 192.168.1.1 (again depending on your Isp router Ip address)
2. follow these instructions to open the static route page. [Route - SAGEMCOM Swan F5366s User Manual [Page 19] | ManualsLib ](https://www.manualslib.com/manual/2042995/Sagemcom-Swan-F5366s.html?page=19#manual)
3. enter the following setting
4. IP address 192.168.200.0 (in your case 192.168.199.0)
5. Gateway IP address 192.168.1.2 (whatever your OpenWRT router Ip address is)
6. Interface IP_BR_LAN (in my case it was br0)
7. Subnet mask 255.255.255.0 (if there is a field for subnet mask)
8. Save and reboot both your ISP modem/router and OpenWrt router.
This will complete step 2.
---------------------------------------------------------------------------------------------------------
**Step 3** 
1. Download and install the latest Winscp from [WinSCP:: Official Site:: Download ](https://winscp.net/eng/download.php)
2. Run WinScp
3. select Scp as protocol.
  * login with your OpenWrt’s ssh credentials e.g
![image](https://user-images.githubusercontent.com/83491212/183281709-5f1765ea-5bf9-422a-b6ee-bd29e00804f5.png)
 * accept the terms etc etc
4. You will be presented with these windows with OpenWRT folders.double click on three dots icons
![image](https://user-images.githubusercontent.com/83491212/183281721-9285da93-7c27-4e8f-b521-5deb04eb9250.png)
you will be presented with
![image](https://user-images.githubusercontent.com/83491212/183281730-924a041b-6d90-41af-9a0e-a45ecdbf6987.png)
5. Navigate to the first file by double-clicking etc and then config. Here you will find all the three files “network” “wireless” and “firewall”
![image](https://user-images.githubusercontent.com/83491212/183281746-53a53a22-65a0-408b-817e-935db2d88ae8.png)
6. double click on the “network” and copy-paste the following code at the end of the files. remember to set defaultroute '0" as mentioned in the GitHub wiki.
```
config interface 'wlan'
        option proto 'dhcp'
        option defaultroute '0'
```
![image](https://user-images.githubusercontent.com/83491212/183281763-3a58b12e-fbfa-46be-a374-abcfef491615.png)
7. repeat the process with “wireless” files by adding the following code to the end of the wireless file. changing according to your inverter’s AP’s Name and BSSID
```
config wifi-iface 'wifinet0'
        option key 'Changeme'
        option ssid 'SUN2000-HV0123456789'
        option device 'radio0'
        option mode 'sta'
        option bssid '94:25:33:12:34:56'
        option encryption 'psk2'
        option network 'wlan'
```
8. Add the following code to the “firewall” file.
```
config zone
        option input 'ACCEPT'
        option name 'sun'
        option output 'ACCEPT'
        option forward 'ACCEPT'
        option masq '1'
        option network 'wlan'

config forwarding
        option dest 'sun'
        option src 'lan'

config forwarding
        option dest 'lan'
        option src 'sun'
```
9. save everything and restart your OpenWrt router.
-----------------------------------------------------------------------------------------------------
**Step 4** 
Firewall rules for the port forwarding

1. Network -- Firewall -- Port forwarding 
2. Setup first rules as follows

* *Name*: Preroute 
* *Protocol:* TCP
* *Source zone*: Lan
* *External port*: 6607
* *Destination zone* : sun(wlan)
* *Internal IP: Select from dropdown*: 192.168.200.1 
![image](https://user-images.githubusercontent.com/83491212/183281772-5e7572f6-36dd-412e-bd74-55ff0f23bdf8.png)

3. Uncheck NAT loopback from the advanced setting tab
![image](https://user-images.githubusercontent.com/83491212/183281779-a755a952-005f-413c-a6e9-a989f6b00f30.png)

4. Add second rule as following 
* *Name* : Post route 
* Protocol TCP
* *Outbound zone* Sun(wlan)
* *Destination address* select from dropdown 192.168.200.1 
* *Destination port*:  6607
* Action: SNAT (there is some debate on this issue but you can use this) 
* Rewrite IP address: select from dropdown 192.168.200.2 (OpenWrt router IP in Huwai inverter subnet) 

![image](https://user-images.githubusercontent.com/83491212/183281785-2c9ad9c1-2939-4cbf-8267-cdfb93f1d296.png)
-----------------------------------------------------------------------------------------------------
**Step five** 
1. Verify connectivity,  Before continuing with the setup of the integration, you should test whether you can ping the inverter on `192.168.200.1`. You can use one of the SSH/Terminal add-ons of HA, or use a terminal in the VSCode add-on. 
2. successful connection will look like this 
![image](https://user-images.githubusercontent.com/83491212/183282221-35b83b36-9be6-4ccd-acca-85d6675b076d.png)
--------------------------------------------------------------------------------------------------------------
Install Wlcrs integration. 

1. [Connect Home Assistant to the WiFi Access Point of your Huawei inverter.](https://github.com/wlcrs/huawei_solar/wiki/Connecting-to-the-inverter#getting-connectivity-between-ha-on-your-home-network-and-the-inverter-ap)
2. Install this integration with HACS, or copy the contents of this repository into the `custom_components/huawei_solar` directory
3. Restart HA
4. Start the configuration flow:
  * [![Start Config Flow](https://camo.githubusercontent.com/637fd24a458765d763a6dced4b312c503f54397bdd9b683584ef8054f305cd7f/68747470733a2f2f6d792e686f6d652d617373697374616e742e696f2f6261646765732f636f6e6669675f666c6f775f73746172742e737667)](https://my.home-assistant.io/redirect/config_flow_start?domain=huawei_solar)
  * Or: Go to `Configuration` -> `Integrations` and click the `+ Add Integration`. Select `Huawei Solar` from the list
5. Choose whether you want to connect via serial or network connection
![image](https://user-images.githubusercontent.com/83491212/183282011-c9a806cf-3760-4a3d-9b50-85dcb41190ac.png)

6. Enter the IP address of your inverter (192.168.200.1). The slave id is typically 0. You should only check the `Advanced: enable parameter configuration` checkbox if you intend to dynamically change your battery settings.
![image](https://user-images.githubusercontent.com/83491212/183282149-2649e19f-9d9e-4310-ad81-2b9b86be08c8.png)

7. When using the `elevate permissions` feature, you might be asked to enter the credentials to the `installer` account in the next step. These are the credentials used to connect to the inverter in the "Device Commissioning" section of the FusionSolar App. The default password is 00000a.

</p>
</details>

**Reply by hirofairlane** [reply]
Thanks a lot for the information published about the setup of Openwrt, but after some hours tinkering, I think I have found a very neat approach.

The way I have configured the data redirection between the inverter SUNXXX subnet and my LAN, is just setting up a SNAT in the Openwrt router that is connected to the SUN2000 wifi and to my LAN. 
Then the host you have to setup in the integration is the ip from the Openwrt that is bridging between networks (AKA Rewrite IP address). In my case 192.168.1.7 that was the LAN IP of the router I have in the basement closer to the inverter.

<img width="466" alt="Screenshot 2022-09-15 171013" src="https://user-images.githubusercontent.com/25512210/190440517-4d7193fb-27e7-41f7-ad25-1a0deeacf20c.png">

in the setup file /etc/config/firewall looks like this:

```
config nat
        option name 'inversor'
        list proto 'tcp'
        option src '*'
        option src_port '6607'
        option dest_ip '192.168.200.1'
        option dest_port '6607'
        option target 'SNAT'
        option snat_ip '192.168.1.7'
```

**Comment by yousaf465** [comment]
It was already discussed in Openwrt forums
https://forum.openwrt.org/t/connect-to-huwaei-sun2000-inverter-using-fiber-modem/104635/25?u=koa

**Reply by hirofairlane** [reply]
Sure, that was the clue I took to make the setup, it should be credited to Mr Trendy and you. I just post it with the aim of confirming that I tested it and worked.

In my opinion it is more simple approach than the one it is currently in the wiki.

**Reply by yousaf465** [reply]
> Mr Trendy did it, I was just asking questions.
